core/Type/Type.php

```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Type;
+
+use Piwik\Cache;
+use Piwik\CacheId;
+use Piwik\Plugin;
+use Piwik\Plugin\Manager as PluginManager;
+use Piwik\Property\PropertySettings;
+
+class Type
+{
+    const ID = '';
+    protected $name = 'General_Property';
+    protected $namePlural = 'General_Properties';
+    protected $description = 'Default type';
+    protected $howToSetupUrl = '';
+
+    public function isType($typeId)
+    {
+        // here we should add some point also check whether id matches any extended ID. Eg if
+        // MetaSites extends Websites, then we expected $metaSite->isType('website') to be true (maybe)
+        return $this->getId() === $typeId;
+    }
+
+    public function getId()
+    {
+        $id = static::ID;
+
+        if (empty($id)) {
+            $message = 'Type %s does not define an ID. Set the ID constant to fix this issue';;
+            throw new \Exception(sprintf($message, get_called_class()));
+        }
+
+        return $id;
+    }
+
+    public function getDescription()
+    {
+        return $this->description;
+    }
+
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    public function getNamePlural()
+    {
+        return $this->namePlural;
+    }
+
+    public function getHowToSetupUrl()
+    {
+        return $this->howToSetupUrl;
+    }
+
+    public function configurePropertySettings(PropertySettings $settings)
+    {
+    }
+
+    /**
+     * @return Type[]
+     */
+    public static function getAllTypes()
+    {
+        $cache    = Cache::getTransientCache();
+        $cacheKey = CacheId::pluginAware('Type_AllTypes');
+
+        if ($cache->contains($cacheKey)) {
+            $types = $cache->fetch($cacheKey);
+        } else {
+            $types = PluginManager::getInstance()->findComponents('Type', '\\Piwik\\Type\\Type');
+            $cache->save($cacheKey, $types);
+        }
+
+        return $types;
+    }
```

Comment:
I do know that and it's a trade-off. Nothing of this is API and it will be changed soon later again anyway in a most likely very different way. Therefore I'd rather add it in the way we did it everywhere so far. And I would most likely add an integration test for this anyway. Having only unit tests for something is not good either. 

> Same for the plugin manager, we can mock the call and the test will be just as good.

A mock will be never as good as testing the "reality" but it comes close yes and I totally agree that it makes sense sometimes but not always.  

> Yes that means the feature is delivered a little sooner, but that also means that we are introducing code that we will have to refactor 

As mentioned we will very soon remove this code again anyway

>  and there's a high chance the integration test will stay an integration test anyway.

That's nothing bad IMO

Anyway, I can quickly introduce a `Type\Manager` or so. Will change the naming of some classes anyway